## Installation

1. If you haven't already installed Python3, get it from [Python.org](https://www.python.org/downloads/)
1. If you haven't already installed Jupyter Notebook, run `python3 -m pip install jupyter`
1. In Terminal, cd to the folder in which you downloaded this file and run `jupyter notebook`. This should open up a page in your web browser that shows all of the files in the current directory, so that you can open this file. You will need to leave this Terminal window up and running and use a different one for the rest of the instructions.
1. Install the Gensim word2vec Python implementation: `pip3 install --upgrade gensim`
1. Get the trained model (1billion_word_vectors.zip) from me via airdrop or flashdrive and put it in the same folder as the ipynb file, the folder in which you are running the jupyter notebook command.
1. Unzip the trained model file. You should now have three files in the folder (if zip created a new folder, mor:
    * 1billion_word_vectors
    * 1billion_word_vectors.syn1neg.npy
    * 1billion_word_vectors.wv.syn0.npy
1. If you didn't install keras last time, install it now
    1. Install the tensorflow machine learning library by typing the following into Terminal:
    `pip3 install --upgrade tensorflow`
    1. Install the keras machine learning library by typing the following into Terminal:
    `pip3 install keras`


## Extra Details -- Do Not Do This
This took awhile, which is why I'm giving you the trained file rather than having you do this. But just in case you're curious, here is how to create the trained model file.
1. Download the corpus of sentences from [http://www.statmt.org/lm-benchmark/1-billion-word-language-modeling-benchmark-r13output.tar.gz](http://www.statmt.org/lm-benchmark/1-billion-word-language-modeling-benchmark-r13output.tar.gz)
1. Unzip and unarchive the file: `tar zxf 1-billion-word-language-modeling-benchmark-r13output.tar.gz` 
1. Run the following Python code:
    ```
    from gensim.models import word2vec
    import os

    corpus_dir = '1-billion-word-language-modeling-benchmark-r13output/training-monolingual.tokenized.shuffled'
    sentences = word2vec.PathLineSentences(corpus_dir)
    model = word2vec.Word2Vec(sentences) # just use all of the default settings for now
    model.save('1billion_word_vectors')
    ```

## Documentation/Sources
* [https://radimrehurek.com/gensim/models/word2vec.html](https://radimrehurek.com/gensim/models/word2vec.html) for more information about how to use gensim word2vec in general
* [https://codekansas.github.io/blog/2016/gensim.html](https://codekansas.github.io/blog/2016/gensim.html) for information about using it to create embedding layers for neural networks.
* [https://machinelearningmastery.com/sequence-classification-lstm-recurrent-neural-networks-python-keras/](https://machinelearningmastery.com/sequence-classification-lstm-recurrent-neural-networks-python-keras/) for information on sequence classification with keras
* [https://blog.keras.io/using-pre-trained-word-embeddings-in-a-keras-model.html](https://blog.keras.io/using-pre-trained-word-embeddings-in-a-keras-model.html) for using pre-trained embeddings with keras (though the syntax they use for the model layers is different than most other tutorials I've seen).
* [https://keras.io/](https://keras.io/) Keras API documentation

### Brief Note on Word Vectors:
The word vector gives the word a location that captures how it relates to other words: i.e. synonyms or words that play similar roles

## Load the trained word vectors

In [1]:
from gensim.models import word2vec

Load the trained model file into memory

In [2]:
wv_model = word2vec.Word2Vec.load('1billion_word_vectors')

Since we do not need to continue training the model, we can save memory by keeping the parts we need (the word vectors themselves) and getting rid of the rest of the model.

In [3]:
wordvec = wv_model.wv #takes only the word vectors from the file
del wv_model

## Exploration of word vectors
Now we can look at some of the relationships between different words.

Like [the gensim documentation](https://radimrehurek.com/gensim/models/word2vec.html), let's start with a famous example: king + woman - man

In [4]:
wordvec.most_similar(positive=['king', 'woman'], negative=['man'])

[('queen', 0.8407387733459473),
 ('monarch', 0.7541723251342773),
 ('prince', 0.7350202798843384),
 ('princess', 0.6969081163406372),
 ('empress', 0.677180290222168),
 ('sultan', 0.6649758815765381),
 ('Chakri', 0.6451101899147034),
 ('goddess', 0.6439394950866699),
 ('ruler', 0.6275453567504883),
 ('kings', 0.6273428201675415)]

Another Word Vector Example: 

In [5]:
wordvec.most_similar(positive=['worker', 'woman'], negative=['man'])

[('nurse', 0.7360476851463318),
 ('caseworker', 0.6824815273284912),
 ('receptionist', 0.6763378381729126),
 ('firefighter', 0.6593405604362488),
 ('midwife', 0.653285026550293),
 ('student', 0.6520328521728516),
 ('janitor', 0.6445921659469604),
 ('roofer', 0.6432132720947266),
 ('labourer', 0.6422997713088989),
 ('technician', 0.6403226852416992)]

In [6]:
wordvec.most_similar(positive=['asian', 'worker'])

[('african', 0.6535460948944092),
 ('labourer', 0.56971675157547),
 ('student', 0.5687010288238525),
 ('butte', 0.5473869442939758),
 ('florist', 0.5461958646774292),
 ('educator', 0.5448765158653259),
 ('british', 0.5441399812698364),
 ('electrician', 0.5435316562652588),
 ('employee', 0.5426956415176392),
 ('gamekeeper', 0.5414238572120667)]

In [7]:
wordvec.doesnt_match(['red', 'purple', 'laptop', 'turquoise', 'ruby']) #finding the word that doesn't match

'laptop'

How far apart are different words?

In [8]:
wordvec.distances('laptop', ['computer', 'phone', 'rabbit']) 
#comparing the difference between laptop and the three words in the array

array([0.205414  , 0.36557418, 0.65974367], dtype=float32)

Let's see what one of these vectors actually looks like.

In [9]:
wordvec['textbook'] #returns a multi-dimensional array (each one represents 100 diff dimensions)
#though there are 100 dimensions shown, there can be 'hidden' vectors that the model can use to locate different
#features that may not be represented here 

array([ 0.50756323, -2.8890731 ,  0.9743826 , -0.60089743, -0.23762947,
       -2.324566  , -0.64634913, -0.66476715, -2.3432739 ,  1.4446437 ,
       -0.15542823,  1.8248576 ,  1.1309539 , -0.21071543, -0.82512087,
       -0.2773584 , -0.1973424 , -0.5337731 ,  2.1143918 ,  1.0673765 ,
       -0.2341243 ,  1.5292411 ,  0.66977274,  1.1214821 , -0.57710004,
       -0.02504024,  0.6074397 ,  0.19416903, -1.1265849 , -0.6618393 ,
        1.7525213 ,  1.6232891 , -0.3886833 , -1.1867149 ,  0.45511633,
        1.4240934 , -0.87929034, -1.8920534 ,  2.6986032 , -0.5277589 ,
        2.1202435 ,  0.62670445,  1.0352231 ,  1.4998924 ,  2.5809426 ,
        0.74698585, -0.07757699, -0.67074645,  1.6887746 , -0.22081567,
        1.2107906 ,  0.16741815,  3.3496742 ,  1.1832954 ,  0.4423463 ,
        0.04771314, -0.14557275, -1.3345221 ,  1.3236852 ,  2.0154989 ,
       -0.6510446 ,  0.21808812, -0.31578887, -1.822629  ,  0.8436349 ,
       -1.1500564 ,  1.24044   , -2.6430037 ,  1.0617311 ,  1.20

What other methods are available to us?

In [10]:
#help(wordvec)

# Exercise: Explore Word Vectors
What other interesting relationship can you find, using the methods used in the examples above or anything you find in the help message?

In [11]:
wordvec.similar_by_word('man')

[('woman', 0.9145064353942871),
 ('teenager', 0.8810972571372986),
 ('boy', 0.8332439661026001),
 ('girl', 0.8199804425239563),
 ('policeman', 0.7857508659362793),
 ('burglar', 0.7623993754386902),
 ('pensioner', 0.7541862726211548),
 ('toddler', 0.7527749538421631),
 ('soldier', 0.7502890825271606),
 ('thief', 0.7498296499252319)]

In [12]:
wordvec.similarity('woman','man')

0.9145063682687546

## Using the word vectors in an embedding layer of a Keras model

In [13]:
from keras.models import Sequential
import numpy

Using TensorFlow backend.
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: compiletime version 3.5 of module 'tensorflow.python.framework.fast_tensor_util' does not match runtime version 3.6
  return f(*args, **kwds)


You may have noticed in the help text for wordvec that it has a built-in method for converting into a Keras embedding layer.

In [14]:
# Returns a Keras 'Embedding' layer with weights set as the wordvec model's learned word embeddings
# embedding is getting the words into a vector space
test_embedding_layer = wordvec.get_keras_embedding()
test_embedding_layer.input_length = 1

In [15]:
embedding_model = Sequential()
embedding_model.add(test_embedding_layer)
#using the embedding model transforms the wordvec into a format understandable for Keras as an input
# the model will take as input an integer matrix of size (batch, input_length).

But how do we actually use this? If you look at the [Keras Embedding Layer documentation](https://keras.io/layers/embeddings/) you might notice that it takes numerical input, not strings. How do we know which number corresponds to a particular word? In addition to having a vector, each word has an index:

In [16]:
wordvec.vocab['python'].index #another way to categorize the vectors

30438

Let's see if we get the same vector from the embedding layer as we get from our word vector object.

In [17]:
wordvec['python'] #it seems to match 

array([-1.1750487e+00,  2.3066440e-04, -6.0706180e-01, -1.1156354e+00,
       -1.0580894e+00, -2.7154784e+00, -3.6140988e+00, -1.0810910e+00,
        1.1234255e+00, -7.7326834e-01, -1.3322397e+00,  9.2905626e-02,
       -2.4488842e+00, -1.7817341e-01, -3.5459950e+00, -1.7320968e+00,
        1.9397168e+00, -6.3734710e-01,  2.3254216e+00, -1.3535864e+00,
       -1.4451812e-01, -2.4297442e+00,  1.5498929e+00,  8.1969726e-01,
        9.0982294e-01, -6.6116208e-01,  3.8905215e-01,  3.3855909e-01,
       -7.5454485e-01, -1.0352553e+00, -2.5936973e+00,  1.2103225e+00,
       -3.0236175e+00,  3.0580134e+00, -3.9140179e+00,  4.0223894e-01,
        1.7356061e+00,  9.0976155e-01,  2.0956397e-02,  2.0190549e+00,
        4.5332021e-01, -1.6634842e+00, -4.8180079e-01,  2.0414692e-01,
       -5.9267312e-01, -1.4182589e+00, -9.7301149e-01,  5.1611459e-01,
        2.0727324e+00,  2.0064230e+00, -7.5027935e-02, -1.1723986e+00,
       -8.6943096e-01,  1.7028141e+00,  2.2190344e+00,  9.3605727e-01,
      

In [18]:
embedding_model.predict(numpy.array([[30438]]))

array([[[-1.1750487e+00,  2.3066440e-04, -6.0706180e-01, -1.1156354e+00,
         -1.0580894e+00, -2.7154784e+00, -3.6140988e+00, -1.0810910e+00,
          1.1234255e+00, -7.7326834e-01, -1.3322397e+00,  9.2905626e-02,
         -2.4488842e+00, -1.7817341e-01, -3.5459950e+00, -1.7320968e+00,
          1.9397168e+00, -6.3734710e-01,  2.3254216e+00, -1.3535864e+00,
         -1.4451812e-01, -2.4297442e+00,  1.5498929e+00,  8.1969726e-01,
          9.0982294e-01, -6.6116208e-01,  3.8905215e-01,  3.3855909e-01,
         -7.5454485e-01, -1.0352553e+00, -2.5936973e+00,  1.2103225e+00,
         -3.0236175e+00,  3.0580134e+00, -3.9140179e+00,  4.0223894e-01,
          1.7356061e+00,  9.0976155e-01,  2.0956397e-02,  2.0190549e+00,
          4.5332021e-01, -1.6634842e+00, -4.8180079e-01,  2.0414692e-01,
         -5.9267312e-01, -1.4182589e+00, -9.7301149e-01,  5.1611459e-01,
          2.0727324e+00,  2.0064230e+00, -7.5027935e-02, -1.1723986e+00,
         -8.6943096e-01,  1.7028141e+00,  2.2190344

Looks good, right? But let's not waste our time when the computer could tell us definitively and quickly:

In [19]:
embedding_model.predict(numpy.array([[wordvec.vocab['python'].index]]))[0][0] == wordvec['python']
#using the embedding_model instead because it is in the proper input format for Keras

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True])

Now we have a way to turn words into word vectors with Keras layers. Yes! Time to get some data.

## The IMDB Dataset
The [IMDB dataset](https://keras.io/datasets/#imdb-movie-reviews-sentiment-classification) consists of movie reviews that have been marked as positive or negative. (There is also a built-in dataset of [Reuters newswires](https://keras.io/datasets/#reuters-newswire-topics-classification) that have been classified by topic.)

In [20]:
from keras.datasets import imdb
(x_train, y_train), (x_test, y_test) = imdb.load_data()

It looks like our labels consists of 0 or 1, which makes sense for positive and negative.

In [21]:
print(y_train[0:9])
#print(max(y_train))
#print(min(y_train))

[1 0 0 1 0 0 1 0 1]


But x is a bit more trouble. The words have already been converted to numbers -- numbers that have nothing to do with the word embeddings we spent time learning!

In [22]:
print(x_train[0]) #relates to word frequency in that data set 

[1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 22665, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 21631, 336, 385, 39, 4, 172, 4536, 1111, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2025, 19, 14, 22, 4, 1920, 4613, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 1247, 4, 22, 17, 515, 17, 12, 16, 626, 18, 19193, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2223, 5244, 16, 480, 66, 3785, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 1415, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 10311, 8, 4, 107, 117, 5952, 15, 256, 4, 31050, 7, 3766, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 12118, 1029, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2071, 56, 26, 141, 6, 194, 7486, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 5535, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 1334, 88, 12, 16, 283, 5, 16, 4472, 113, 103, 32, 15, 16, 5345, 19, 178, 32]


In [23]:
#can retreive words by using the built-in keras functions 
#help(imdb)

In [24]:
imdb_offset = 3
imdb_map = dict((index + imdb_offset, word) for (word, index) in imdb.get_word_index().items())
imdb_map[0] = 'PADDING' #to keep same input matrix size - for anything below input size
imdb_map[1] = 'START' #to show beginning of review
imdb_map[2] = 'UNKNOWN'#if you only want the most common words, it will assign "2" to whatever you don't want

Reference [this stack overflow post](https://stackoverflow.com/questions/42821330/restore-original-text-from-keras-s-imdb-dataset) for more info on the indeces, otherwise the imdb_map would return gibberish

In [25]:
reviewString = ' '.join([imdb_map[word_index] for word_index in x_train[0]])

## Train our IMDB word vectors

In [26]:
#take into account the offset of the padding
train_sentences = [['PADDING'] + [imdb_map[word_index] for word_index in review] for review in x_train]
test_sentences = [['PADDING'] + [imdb_map[word_index] for word_index in review] for review in x_test]

In [27]:
imdb_wv_model = word2vec.Word2Vec(train_sentences + test_sentences + ['UNKNOWN'], min_count=1)
#setting model to the algorithm, treating each review as one sentence 
#word2vec = algorithm for coming up with word vectors per sentence

In [28]:
imdb_wordvec = imdb_wv_model.wv
del imdb_wv_model #remove anything besides the word vectors 
#we don't need the other stuff if you're not going to keep training 

## Process the dataset
We want all inputs the same length (we'll see how to do variable-length later). This means we need to choose a maximum length for the review, cutting off longer ones and adding padding to shorter ones. What should we make the length? Let's understand our data.

In [29]:
lengths = [len(review) for review in x_train + x_test]
print('Longest review: {} Shortest review: {}'.format(max(lengths), min(lengths)))


Longest review: 2697 Shortest review: 70


2697 words! Wow. Well, let's see how many reviews would get cut off at a particular cutoff.

In [30]:
cutoff = 500
print('{} reviews out of {} are over {}.'.format(
    sum([1 for length in lengths if length > cutoff]), 
    len(lengths), 
    cutoff))

8485 reviews out of 25000 are over 500.


In [31]:
from keras.preprocessing import sequence 
#sequence is about the inputs
#Transform a list of num_samples sequences (lists of scalars) into 
#a 2D Numpy array of shape  (num_samples, num_timesteps)
#num_timesteps is either the maxlen argument if provided, or the length of the longest sequence otherwise
x_train_padded = sequence.pad_sequences(x_train, maxlen=cutoff)
x_test_padded = sequence.pad_sequences(x_test, maxlen=cutoff)

## Classification without using the pre-trained word vectors

In [32]:
#from keras.models import Sequential
#sequential is about the model 
from keras.layers import Embedding, Conv1D, Dense, Flatten

Model definition.

In [33]:
no_vector_model = Sequential()
#input_dim = length of vocab (num of unique words)
no_vector_model.add(Embedding(input_dim=len(imdb_map), output_dim=100, input_length=cutoff))
#5 layers of convo network that can look for 5 different features
no_vector_model.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
no_vector_model.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
no_vector_model.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
no_vector_model.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
no_vector_model.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
no_vector_model.add(Flatten())
no_vector_model.add(Dense(units=128, activation='relu'))
no_vector_model.add(Dense(units=1, activation='sigmoid')) # because at the end, we want one yes/no answer
no_vector_model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['binary_accuracy'])

Train the model. __This takes awhile. You might not want to re-run it.__

In [34]:
no_vector_model.fit(x_train_padded, y_train, epochs=1, batch_size=64)

Epoch 1/1
25000/25000 [==============================] - 153s 6ms/step - loss: 0.4204 - binary_accuracy: 0.7740


Assess the model. __This takes awhile. You might not want to re-run it.__

In [35]:
#separating tests and training data to make sure we did not over fit the model
no_vector_scores = no_vector_model.evaluate(x_test_padded, y_test)
print('loss: {} accuracy: {}'.format(*no_vector_scores))

25000/25000 [==============================] - 40s 2ms/step
loss: 0.3238655872821808 accuracy: 0.86388


It seems that the training data had an accuracy of around .777 while the testing data had an accuracy of around .867

# Exercise: Use the word vectors in a full model
Using the knowledge about how the imdb dataset and the keras embedding layer represent words, as detailed above, change the model in the three cells above this one to use the word vectors rather than an embedding that keras learns as it goes along.

You'll need to swap out the embedding layer and feed in different training data.

In [36]:
imdb_embedding_layer = imdb_wordvec.get_keras_embedding()
imdb_embedding_layer.input_length = cutoff

In [37]:
no_vector_model1 = Sequential()
no_vector_model1.add(imdb_embedding_layer)
#5 layers of convo network that can look for 5 different features
no_vector_model1.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
no_vector_model1.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
no_vector_model1.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
no_vector_model1.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
no_vector_model1.add(Conv1D(filters=32, kernel_size=5, activation='relu'))
no_vector_model1.add(Flatten())
no_vector_model1.add(Dense(units=128, activation='relu'))
no_vector_model1.add(Dense(units=1, activation='sigmoid')) # because at the end, we want one yes/no answer
no_vector_model1.compile(loss='binary_crossentropy', optimizer='adam', metrics=['binary_accuracy'])

In [38]:
word_index_array=[]#this array will hold each review, which consists of an array of words
for review in x_train_padded:
    word_index_arrayReview=[]#this is an array of the words in each review
    for word_imdb_num in review:
        word = imdb_map[word_imdb_num]#getting the word that corresponds w/ num in imdb dataset
        word_index = imdb_wordvec.vocab[word].index #getting the index for each word in the array
        word_index_arrayReview.append(word_index)
    word_index_array.append(word_index_arrayReview)

In [39]:
no_vector_model1.fit(word_index_array, y_train, epochs=1, batch_size=64)

Epoch 1/1
25000/25000 [==============================] - 106s 4ms/step - loss: 0.4913 - binary_accuracy: 0.7441


This is an example of the output:
Epoch 1/1
25000/25000 [==============================] - 106s 4ms/step - loss: 0.4913 - binary_accuracy: 0.7441

# Questions (Optional)
1. How does the performance of the two models (with and without trained word vectors) compare?
2. Why use one-dimensional convolutional layers for this problem? How do they help?